From 87ba6a3323e49d533d19b4b71be909dcc16185ca Mon Sep 17 00:00:00 2001
From: Amy Zhang <Amy.Zhang@amd.com>
Date: Fri, 16 Sep 2016 14:51:53 -0400
Subject: [PATCH 1090/1722] drm/amd/dal: Handle PSR in notify vsync

- Check freesync state before calling PSR
- Called PSR enable/disable in notify vsync interrupt state
- Disabled PSR during mode change

Change-Id: Ic56c47b19f6a8f7d81ce23b9fb0342deb934af80
Signed-off-by: Amy Zhang <Amy.Zhang@amd.com>
Acked-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 .../gpu/drm/amd/dal/modules/freesync/freesync.c    | 28 ++++++++++++++++++++++
 drivers/gpu/drm/amd/dal/modules/inc/mod_freesync.h |  4 ++++
 2 files changed, 32 insertions(+)

diff --git a/drivers/gpu/drm/amd/dal/modules/freesync/freesync.c b/drivers/gpu/drm/amd/dal/modules/freesync/freesync.c
index bc38eaf9..7b89dfa 100644
--- a/drivers/gpu/drm/amd/dal/modules/freesync/freesync.c
+++ b/drivers/gpu/drm/amd/dal/modules/freesync/freesync.c
@@ -835,6 +835,34 @@ void mod_freesync_update_state(struct mod_freesync *mod_freesync,
 		set_freesync_on_streams(core_freesync, streams, num_streams);
 }
 
+
+bool mod_freesync_get_state(struct mod_freesync *mod_freesync,
+		const struct dc_sink *sink,
+		struct mod_freesync_params *freesync_params)
+{
+	struct core_freesync *core_freesync =
+				MOD_FREESYNC_TO_CORE(mod_freesync);
+
+	unsigned int index = map_index_from_sink(core_freesync, sink);
+
+	if (core_freesync->map[index].state.fullscreen) {
+		freesync_params->state = FREESYNC_STATE_FULLSCREEN;
+		freesync_params->enable = true;
+	} else if (core_freesync->map[index].state.static_screen) {
+		freesync_params->state = FREESYNC_STATE_STATIC_SCREEN;
+		freesync_params->enable = true;
+	} else if (core_freesync->map[index].state.video) {
+		freesync_params->state = FREESYNC_STATE_VIDEO;
+		freesync_params->enable = true;
+	} else
+		freesync_params->enable = false;
+
+	freesync_params->update_duration_in_ns =
+		core_freesync->map[index].state.time.update_duration_in_ns;
+
+	return true;
+}
+
 bool mod_freesync_get_freesync_caps(struct mod_freesync *mod_freesync,
 		const struct dc_sink *sink, struct mod_freesync_caps *caps)
 {
diff --git a/drivers/gpu/drm/amd/dal/modules/inc/mod_freesync.h b/drivers/gpu/drm/amd/dal/modules/inc/mod_freesync.h
index f22be6c..5390b19 100644
--- a/drivers/gpu/drm/amd/dal/modules/inc/mod_freesync.h
+++ b/drivers/gpu/drm/amd/dal/modules/inc/mod_freesync.h
@@ -106,6 +106,10 @@ void mod_freesync_update_state(struct mod_freesync *mod_freesync,
 		const struct dc_stream **streams, int num_streams,
 		struct mod_freesync_params *freesync_params);
 
+bool mod_freesync_get_state(struct mod_freesync *mod_freesync,
+		const struct dc_sink *sink,
+		struct mod_freesync_params *freesync_params);
+
 bool mod_freesync_get_freesync_caps(struct mod_freesync *mod_freesync,
 		const struct dc_sink *sink, struct mod_freesync_caps *caps);
 
-- 
2.7.4

